### PR TITLE
samples: nrf9160: modem_shell: link: updates

### PR DESCRIPTION
**Pull request by jhirsi**
This PR adds following new functionalities and updates to mosh `link` command:
- `link ncellmeas`: new optional search types and prints additionally TA measurement time as a results
- `link settings`: option to reset modem settings as their defaults
- config: default system mode changed to LTE-M.

### CODE REVIEW

File: samples/nrf9160/modem_shell/src/link/link_settings.c [hunk 1]
```diff
@@ -679,6 +681,19 @@ void link_sett_defaults_set(void)
...
 }
 
+void link_sett_modem_factory_reset(enum link_sett_modem_reset_type type)
+{
+	int err;
+
+	/* Reset modem factory settings by type */
+	err = nrf_modem_at_printf("AT%%XFACTORYRESET=%d", type);
+	if (err) {
+		mosh_error("Resetting modem factory settings failed, err %d\n", err);
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/link/link_settings.c:691]:
```suggestion
		mosh_error("Resetting modem factory settings failed, err %d", err);
```

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 2]
```diff
@@ -234,11 +238,24 @@ static const char link_rsrp_usage_str[] =
 	"  -u, --unsubscribe,  Unsubscribe for RSRP info\n";
 
 static const char link_ncellmeas_usage_str[] =
-	"Usage: link ncellmeas --single | --continuous |--cancel\n"
+	"Usage: link ncellmeas --single | --continuous |--cancel [<search_type>]\n"
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:241]: At least to me this notation is a bit confusing, it looks as if `search_type` would be an option for `--cancel`. Also a space is missing before `--cancel`. How about this?

```suggestion
	"Usage: link ncellmeas --single | --continuous | --cancel | [<search_type>]\n"
```

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 3]
```diff
@@ -234,11 +238,24 @@ static const char link_rsrp_usage_str[] =
...
 	"Options:\n"
-	"      --single,     Start a neighbor cell measurement and report result\n"
-	"      --continuous, Start continuous neighbor cell measurement mode and report result\n"
-	"      --cancel,     Cancel/Stop neighbor cell measurement if still ongoing\n";
+	"      --single,       Start a neighbor cell measurement and report result\n"
+	"      --continuous,   Start continuous neighbor cell measurement mode and report result\n"
+	"      --cancel,       Cancel/Stop neighbor cell measurement if still ongoing\n"
+	"Optional options:\n"
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:246]: I guess an "option" is optional without being an "optional option". :) Also these are valid for single and continuous measurements, but not for cancel.
```suggestion
	"Search type options, can be used when starting measurements:\n"
```

**Review by tokangas — APPROVED** [review]
Looks good to me, only some minor comments.

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 3 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c]: Changed.

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 2 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c]: yes, that is how it should be.

File: samples/nrf9160/modem_shell/src/link/link_settings.c (hunk 1 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_settings.c]: Changed.

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 4]
```diff
@@ -234,11 +238,24 @@ static const char link_rsrp_usage_str[] =
...
+	"                        based on previous cell history. For modem firmware\n"
+	"                        versions < 1.3.1, this is the only valid option.\n"
+	"      --type_ext_light, Search type: extended light.  The modem starts with the same\n"
+	"                        search method as the default type. If needed, it continues to\n"
+	"                        search by measuring the radio conditions and makes assumptions on\n"
+	"                        where networks might be deployed, i.e. a light search.\n"
+	"      --type_ext_comp,  Search type: extended complete.  The modem follows the same\n"
+	"                        procedure as for type_ext_light, but will continue to perform\n"
+	"                        a complete search instead of a light search, and the search is\n"
+	"                        performed for all supported bands.\n";
```
Comment (trantanen) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:258]: I'm seeing now after having written this that TommiK also commented it. Below is still my original comment which we should think about.

The style in which search type is defined doesn't seem to fit well into the usage format (at least couldn't find where we would have used this) although I have no problems understanding it.
I think we have two options for a more normal usage style.

1. `--search_type=ext_light` type of style where values could be `default`, `ext_light`, `ext_comp`.
2. Keep two existing ones (`--type_ext_light` and `--type_ext_comp`) and add `type_default` basically just to have a place where to document the default behavior. Not sure whether this makes sense.

**Review by rlubos — APPROVED** [review]

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 5]
```diff
@@ -107,10 +109,12 @@ static const char link_usage_str[] =
...
 static const char link_settings_usage_str[] =
-	"Usage: link settings --read | --reset\n"
+	"Usage: link settings --read | --reset | --mreset_all | --mreset_user\n"
 	"Options:\n"
-	"  -r, --read,   Read and print current persistent settings\n"
-	"      --reset,  Reset all persistent settings as their defaults\n";
+	"  -r, --read,        Read and print current persistent settings\n"
+	"      --reset,       Reset all persistent settings as their defaults\n"
+	"      --mreset_all,  Reset all modem settings as their defaults\n"
+	"      --mreset_user, Reset modem user configurable settings as their defaults\n";
```
Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:117]: "Reset all persistent settings to defaults" and same for other lines.

**Review by hiltunent — APPROVED** [review]

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 5 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:117]: Original `--reset` hook is having: "Reset all persistent settings as their defaults"
I recall that it was pretty much like you proposed but this current was something that somebody wanted it to be in some previous review and these new ones are following that ;-)

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 4 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:258]: After doing TommiK's proposal for usage, I think it is simpler and fine as it is.

File: samples/nrf9160/modem_shell/src/link/link.c [hunk 6]
```diff
@@ -274,13 +277,15 @@ void link_ind_handler(const struct lte_lc_evt *const evt)
 
 			mosh_print("  Current cell:");
 			mosh_print(
-				"    ID %d, phy ID %d, MCC %d MNC %d, RSRP %d : %ddBm, RSRQ %d, TAC %d, earfcn %d, meas time %lld, TA %s",
+				"    ID %d, phy ID %d, MCC %d MNC %d, RSRP %d : %ddBm, RSRQ %d, TAC %d, earfcn %d, meas time %lld\n"
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link.c:280]: Was thinking the same but It is now actually shorter as it was and wanted to split just TA stuff to own line.

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 4 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link_shell.c:258]: This is up to you but having an optional parameter which is likely not used most of the time, option #1 would be cleanest from the usage documentation point of view (similarly to, for example, `--wifi_service`).

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 5 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c]: Changed.

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 4 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c]: Changed. Please, check. The changes are in a separate commit for now.

File: samples/nrf9160/modem_shell/src/link/link_shell.c [hunk 7]
```diff
@@ -238,24 +238,27 @@ static const char link_rsrp_usage_str[] =
...
-	"                        versions < 1.3.1, this is the only valid option.\n"
-	"      --type_ext_light, Search type: extended light.  The modem starts with the same\n"
-	"                        search method as the default type. If needed, it continues to\n"
-	"                        search by measuring the radio conditions and makes assumptions on\n"
-	"                        where networks might be deployed, i.e. a light search.\n"
-	"      --type_ext_comp,  Search type: extended complete.  The modem follows the same\n"
-	"                        procedure as for type_ext_light, but will continue to perform\n"
-	"                        a complete search instead of a light search, and the search is\n"
-	"                        performed for all supported bands.\n";
+	"      --search_type,  Used Wi-Fi positioning service:\n"
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/link/link_shell.c:246]: "Used Wi-Fi positioning service" doesn't look right here. :)

Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/link/link_shell.c]: ;-) changed

File: samples/nrf9160/modem_shell/src/link/link_shell.c (hunk 4 above)
Comment (trantanen) [reply samples/nrf9160/modem_shell/src/link/link_shell.c]: This is good

**Review by trantanen — APPROVED** [review]

**Pull request merged**